components/slack/new-message-in-conversation.js

```diff
@@ -0,0 +1,109 @@
+const slack = require('https://github.com/PipedreamHQ/pipedream/components/slack/slack.app.js')
+
+module.exports = {
+  name: "Slack - New Message In Conversation(s)",
+  dedupe: "unique",
+  props: {
+    slack,
+    conversations: {
+      type: "string[]",
+      label: "Conversations",
+      description: "Conversations you'd like to receive notifications for.",
+      optional: true,
+      async options({ prevContext }) {
+        let { types, cursor, userNames } = prevContext
+        if (types == null) {
+          scopes = await this.slack.scopes()
+          types = ["public_channel"]
+          if (scopes.includes("groups:read")) {
+            types.push("private_channel")
+          }
+          if (scopes.includes("mpim:read")) {
+            types.push("mpim")
+          }
+          if (scopes.includes("im:read")) {
+            types.push("im")
+            // TODO use paging
+            userNames = {}
+            for (const user of (await this.slack.users()).users) {
+              userNames[user.id] = user.name
+            }
+          }
+        }
+        const resp = await this.slack.availableConversations(types.join(), cursor)
+        return {
+          options: resp.conversations.map((c) => {
+            if (c.is_im) {
+              return { label: `Direct messaging with: @${userNames[c.user]}`, value: c.id }
+            } else if (c.is_mpim) {
+              return { label: c.purpose.value, value: c.id }
+            } else {
+              return { label: `${c.is_private ? "Private" : "Public"} channel: ${c.name}`, value: c.id }
+            }
+          }),
+          context: { types, cursor: resp.cursor, userNames },
+        }
+      },
+    },
+    slackApphook: {
+      type: "$.interface.apphook",
+      appProp: "slack",
+      async eventNames() {
+        return this.conversations || []
+      },
+    },
+    ignoreMyself: {
+      type: "boolean",
+      label: "Ignore myself",
+      description: "Ignore messages from me",
+      default: true,
+    },
+    resolveNames: {
+      type: "boolean",
+      label: "Resolve names",
+      description: "Resolve user and channel names (incurs extra API calls)",
+      default: false,
+    }
+  },
+  async run(event) {
+    if (event.subtype != null) {
+      // This source is designed to just emit an event for each new message received.
+      // Due to inconsistencies with the shape of message_changed and message_deleted
+      // events, we are ignoring them for now. If you want to handle these types of
+      // events, feel free to change this code!!
+      console.log("Ignoring message with subtype.")
+      return
+    }
+    if (this.ignoreMyself && event.user == this.slack.mySlackId()) {
+      return
+    }
+    if (this.resolveNames) {
```

Comment:
would it make sense to use `db` to cache resolved names? it's bit more code but possibly much, much less wasteful

Reply:
although if cache could be out of date, might be a class of issues we'd rather avoid... for high volume slacks though would rather at least limit how often we end up resolving the same thing

Reply:
do you see this as a blocker? what would you propose? refresh once a day?